File: src/jianzhi21.cpp
class Solution {
public:
    vector<int> exchange(vector<int>& nums) {
        //刚刚做困难题做的我气，做个简单泄愤
        //空数组，只有一个元素的数组
        if(nums.size() == 0 || nums.size() == 1){
            return nums;
        }
        else{
            //双指针，一个从前往后，一个从后往前
            //当前指针指向偶数，后指针指向奇数时交换数字
            //时间击败：44.14%,空间击败：46.33%
            //谨记-比较运算符的优先级高于位运算符
            int pPre = 0;
            int pLast = nums.size() - 1;
            while(pPre <= pLast){
                //移动，让前指针去找偶数
                while(pPre <= pLast && (nums[pPre] & 0x1) != 0){
                    pPre ++;    //停下时指向偶数
                }
                while(pPre <= pLast && (nums[pLast] & 0x1) == 0){
                    pLast --;
                }
                //cout<<pPre<<" "<<nums[pPre]<<" "<<pLast<<" "<<nums[pLast]<<endl;
                //停下时，交换
                if(pPre <= pLast){
                    //int的构造和赋值的开销区别不大，因此可以在内部声明
                    int temp = nums[pPre];
                    nums[pPre] = nums[pLast];
                    nums[pLast] = temp;
                }
            }
            return nums;
        }
    }
};

File: src/jianzhi21.py
class Solution:
    def exchange(self, nums: List[int]) -> List[int]:
        #开启今日的干翻蚂蚁金服
        #一个前指针，一个后指针
        #当前偶后奇时交换，然后两个指针推进
        #如果前偶后偶，后面推进
        #如果前奇后奇，前面退进
        #如果前奇后偶，都推进
        #边界处理
        #时间复杂度On，空间O1
        #十分钟，一次提交，时间击败-78.03%，空间击败-26.90%
        length = len(nums)
        if length <= 1:
            return nums
        left, right = 0, length - 1
        while left < right:
            leftResult = (nums[left] & 0x1) #奇1偶0
            rightResult = (nums[right] & 0x1)
            if leftResult == 0 and rightResult == 1:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1
            elif leftResult == 0 and rightResult == 0:
                right -= 1
            elif leftResult == 1 and rightResult == 1:
                left += 1
            else:
                left += 1
                right -= 1
        return nums
